Librerie/libpdel-0.6.1/pdel/structs/type/structs_type_array.3
.\" @COPYRIGHT@
.\"
.\" Author: Archie Cobbs <archie@freebsd.org>
.\"
.\" $Id: structs_type_array.3 901 2004-06-02 17:24:39Z archie $
.\"
.Dd April 22, 2002
.Dt STRUCTS_TYPE_ARRAY 3
.Os
.Sh NAME
.Nm STRUCTS_ARRAY_TYPE ,
.Nm STRUCTS_FIXEDARRAY_TYPE
.Nd structs types for arrays
.Sh LIBRARY
PDEL Library (libpdel, \-lpdel)
.Sh SYNOPSIS
.In sys/types.h
.In pdel/structs/structs.h
.In pdel/structs/type/array.h
.Fn DEFINE_STRUCTS_ARRAY struct_name elem_ctype
.Fn STRUCTS_ARRAY_TYPE elem_stype mtype xml_tag
.Fn STRUCTS_FIXEDARRAY_TYPE elem_stype elem_size array_len xml_tag
.Ft int
.Fn structs_array_length "const struct structs_type *type" "const char *name" "const void *data"
.Ft int
.Fn structs_array_reset "const struct structs_type *type" "const char *name" "void *data"
.Ft int
.Fn structs_array_insert "const struct structs_type *type" "const char *name" "u_int index" "void *data"
.Ft int
.Fn structs_array_delete "const struct structs_type *type" "const char *name" "u_int index" "void *data"
.Ft int
.Fn structs_array_prep "const struct structs_type *type" "const char *name" "void *data"
.Sh DESCRIPTION
.Ss Macros
The
.Fn STRUCTS_ARRAY_TYPE
and
.Fn STRUCTS_FIXEDARRAY_TYPE
macros define a
.Xr structs 3
type (i.e., a
.Dv "struct structs_type" )
for describing variable and fixed length arrays, respectively.
In both macros,
.Fa elem_stype
is a pointer to the structs type describing the array elements, and
.Fa xml_tag
is an XML tag (ASCII string) used to tag individual array elements
when the array is expressed in XML.
.Pp
For
.Fn STRUCTS_ARRAY_TYPE ,
the described data structure must be a
.Dv "struct structs_array" :
.Pp
.Bd -literal -offset 0n
    struct structs_array {
	u_int   length;       /* number of elements */
	void    *elems;       /* elements */
    };
.Ed
.Pp
The
.Fa length
field contains the number of elements in the array.
The array itself is pointed to by
.Fa elems ,
which must be cast to the appropriate type.
.Fa mtype
is the
.Xr typed_mem 3
type used to dynamically allocate the array.
.Pp
The elements of the array are subfields whose names are their index
in the array.
For example, if a subfield of a data structure named
.Dq foo.bar
has an array type, then the elements of the array are named
.Dq foo.bar.0,
.Dq foo.bar.1,
etc.
.Pp
As a special case,
.Dq foo.bar.length
is a read-only virtual subfield of type
.Xr structs_type_uint 3
that returns the length of the array.
This
.Dq length
field does not appear in the output of
.Xr structs_xml_output 3
or
.Xr structs_traverse 3 .
.Pp
To define a structure equivalent to a
.Li "struct structs_array"
that declares
.Fa elems
to have the correct C type for the array elements, use the
.Fn DEFINE_STRUCTS_ARRAY
macro, where
.Fa struct_name
is the name of the structure (or empty if no name is desired) and
.Fa elem_ctype
is the C type of a single element.
Then the
.Fa elems
field will be declared to have type
.Fa "elem_ctype *" .
.Pp
.Fn STRUCTS_FIXEDARRAY_TYPE
defines a structs type for an array that has a fixed length specified by
.Fa array_len .
The described data structure is simply
.Fa array_len
consecutive instances of
.Fa elem_stype ,
each of which must have size
.Fa elem_size .
In other words,
.Fa elem_size
must be equal to
.Fa "elem_stype->size" .
.Ss Functions
The following functions are included to facilitate handling variable
length arrays.
In all cases,
.Fa type
is the
.Xr structs 3
type for the data structure pointed to by
.Fa data ,
and
.Fa name
indicates the variable length array sub-field of
.Fa data .
If
.Fa name
is
.Dv NULL
or the empty string, then
.Fa data
itself is the variable length array.
.Pp
.Fn structs_array_length
returns the length of the array.
.Pp
.Fn structs_array_reset
resets the array to zero length.
Any existing elements are automatically freed.
.Pp
.Fn structs_array_insert
inserts a new element into the array at position
.Fa index ,
which must be between zero and the length of the array, inclusive.
The new element is automatically initialized to the default value
for its type.
.Pp
.Fn structs_array_delete
frees and removes the element at position
.Fa index ,
which must be greater than or equal to zero and strictly less than the
length of the array.
.Pp
.Fn structs_array_prep
can be used when filling in an array that is initially empty
by consecutively setting each element.
The prefixes of
.Fa name
are taken in order from shortest to longest (i.e.,
.Fa name
itself).
For each prefix that corresponds to an array element, the element
index is examined.
If the index is less than the length of the array, nothing happens.
If the index is greater than the length of the array, an error occurs.
Otherwise, a new element is added to the end of the array.
Therefore, if
.Fn structs_array_prep
is invoked before setting each leaf element (such as returned by
.Xr structs_traverse 3) ,
then the extension of any internal arrays will happen automatically.
.Sh SEE ALSO
.Xr libpdel 3 ,
.Xr structs 3 ,
.Xr structs_type 3 ,
.Xr typed_mem 3
.Sh EXAMPLES
.Bd -literal -offset 0n

/* Define my variable length array of IP addresses structure */
DEFINE_STRUCTS_ARRAY(ip_array, struct in_addr);

/* My variable length array of IP addresses */
static struct ip_array variable_ip_array;

/* Structs type describing "variable_ip_array": >= 0 IP addresses */
static const struct structs_type vip_array_type
    = STRUCTS_ARRAY_TYPE(&structs_type_ip4, "vip_array_memory", "ip");

#define FIXED_ARRAY_LEN  12

/* My fixed length array of IP addresses */
static struct in_addr fixed_ip_array[FIXED_ARRAY_LEN];

/* Structs type describing "fixed_ip_array": 12 IP addresses */
static const struct structs_type vip_array_type
    = STRUCTS_FIXEDARRAY_TYPE(&structs_type_ip4,
	sizeof(struct in_addr), FIXED_ARRAY_LEN, "ip");
.Ed
.Sh HISTORY
The PDEL library was developed at Packet Design, LLC.
.Dv "http://www.packetdesign.com/"
.Sh AUTHORS
.An Archie Cobbs Aq archie@freebsd.org


Librerie/libpdel-0.6.1/pdel/structs/type/structs_type_array.c

/*
 * @COPYRIGHT@
 *
 * Author: Archie Cobbs <archie@freebsd.org>
 */

#include <sys/types.h>
#include <sys/param.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <errno.h>

#include "structs/structs.h"
#include "structs/type/array.h"
#include "util/typed_mem.h"

#define NUM_BYTES(x)	(((x) + 7) / 8)

/*********************************************************************
			ARRAY TYPES
*********************************************************************/

int
structs_array_copy(const struct structs_type *type,
	const void *from, void *to)
{
	const struct structs_type *const etype = type->args[0].v;
	const char *mtype = type->args[1].v;
	const struct structs_array *const fary = from;
	struct structs_array *const tary = to;
	int errno_save;
	u_int i;

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Allocate a new array */
	memset(tary, 0, sizeof(*tary));
	if ((tary->elems = MALLOC(mtype,
	    fary->length * etype->size)) == NULL)
		return (-1);

	/* Copy elements into it */
	for (i = 0; i < fary->length; i++) {
		const void *const from_elem
		    = (char *)fary->elems + (i * etype->size);
		void *const to_elem = (char *)tary->elems + (i * etype->size);

		if ((*etype->copy)(etype, from_elem, to_elem) == -1)
			break;
		tary->length++;
	}

	/* If there was a failure, undo the half completed job */
	if (i < fary->length) {
		errno_save = errno;
		while (i-- > 0) {
			(*etype->uninit)(etype,
			    (char *)tary->elems + (i * etype->size));
		}
		FREE(mtype, tary->elems);
		memset(tary, 0, sizeof(*tary));
		errno = errno_save;
		return (-1);
	}

	/* Done */
	return (0);
}

int
structs_array_equal(const struct structs_type *type,
	const void *v1, const void *v2)
{
	const struct structs_type *const etype = type->args[0].v;
	const struct structs_array *const ary1 = v1;
	const struct structs_array *const ary2 = v2;
	u_int i;

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY)
		return (0);

	/* Check array lengths first */
	if (ary1->length != ary2->length)
		return (0);

	/* Now compare individual elements */
	for (i = 0;
	    i < ary1->length && (*etype->equal)(etype,
		(char *)ary1->elems + (i * etype->size),
		(char *)ary2->elems + (i * etype->size));
	    i++);
	return (i == ary1->length);
}

int
structs_array_encode(const struct structs_type *type, const char *mtype,
	struct structs_data *code, const void *data)
{
	const struct structs_type *const etype = type->args[0].v;
	const struct structs_array *const ary = data;
	const u_int bitslen = NUM_BYTES(ary->length);
	struct structs_data *ecodes;
	u_int32_t elength;
	u_char *bits;
	void *delem;
	u_int tlen;
	int r = -1;
	u_int i;

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Get the default value for an element */
	if ((delem = MALLOC(TYPED_MEM_TEMP, etype->size)) == NULL)
		return (-1);
	if (structs_init(etype, NULL, delem) == -1)
		goto fail1;

	/* Create bit array. Each bit indicates an element that is present. */
	if ((bits = MALLOC(TYPED_MEM_TEMP, bitslen)) == NULL)
		goto fail2;
	memset(bits, 0, bitslen);
	tlen = 4 + bitslen;			/* length word + bits array */

	/* Create array of individual encodings, one per element */
	if ((ecodes = MALLOC(TYPED_MEM_TEMP,
	    ary->length * sizeof(*ecodes))) == NULL)
		goto fail3;
	for (i = 0; i < ary->length; i++) {
		const void *const elem = (char *)ary->elems + (i * etype->size);
		struct structs_data *const ecode = &ecodes[i];

		/* Check for default value, leave out if same as */
		if ((*etype->equal)(etype, elem, delem) == 1) {
			memset(ecode, 0, sizeof(*ecode));
			continue;
		}
		bits[i / 8] |= (1 << (i % 8));

		/* Encode element */
		if ((*etype->encode)(etype, TYPED_MEM_TEMP, ecode, elem) == -1)
			goto fail4;
		tlen += ecode->length;
	}

	/* Allocate final encoded region */
	if ((code->data = MALLOC(mtype, tlen)) == NULL)
		goto fail4;

	/* Copy array length */
	elength = htonl(ary->length);
	memcpy(code->data, &elength, 4);
	code->length = 4;

	/* Copy bits array */
	memcpy(code->data + code->length, bits, bitslen);
	code->length += bitslen;

	/* Copy encoded elements */
	for (i = 0; i < ary->length; i++) {
		struct structs_data *const ecode = &ecodes[i];

		memcpy(code->data + code->length, ecode->data, ecode->length);
		code->length += ecode->length;
	}

	/* OK */
	r = 0;

	/* Clean up and exit */
fail4:	while (i-- > 0)
		FREE(TYPED_MEM_TEMP, ecodes[i].data);
	FREE(TYPED_MEM_TEMP, ecodes);
fail3:	FREE(TYPED_MEM_TEMP, bits);
fail2:	structs_free(etype, NULL, delem);
fail1:	FREE(TYPED_MEM_TEMP, delem);
	return (r);
}

int
structs_array_decode(const struct structs_type *type, const u_char *code,
	size_t cmax, void *data, char *ebuf, size_t emax)
{
	const struct structs_type *const etype = type->args[0].v;
	const char *const mtype = type->args[1].v;
	struct structs_array *const ary = data;
	const u_char *bits;
	u_int32_t elength;
	u_int bitslen;
	int clen;
	u_int i;

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Get number of elements */
	if (cmax < 4)
		goto truncated;
	memcpy(&elength, code, 4);
	ary->length = ntohl(elength);
	code += 4;
	cmax -= 4;
	clen = 4;

	/* Get bits array */
	bitslen = NUM_BYTES(ary->length);
	if (cmax < bitslen) {
truncated:	strlcpy(ebuf, "encoded array is truncated", emax);
		errno = EINVAL;
		return (-1);
	}
	bits = code;
	code += bitslen;
	cmax -= bitslen;
	clen += bitslen;

	/* Allocate array elements */
	if ((ary->elems = MALLOC(mtype, ary->length * etype->size)) == NULL)
		return (-1);

	/* Decode elements */
	for (i = 0; i < ary->length; i++) {
		void *const edata = (char *)ary->elems + (i * etype->size);
		int eclen;

		/* If element not present, assign it the default value */
		if ((bits[i / 8] & (1 << (i % 8))) == 0) {
			if (structs_init(etype, NULL, edata) == -1)
				goto fail;
			continue;
		}

		/* Decode element */
		if ((eclen = (*etype->decode)(etype,
		    code, cmax, edata, ebuf, emax)) == -1)
			goto fail;

		/* Go to next encoded element */
		code += eclen;
		cmax -= eclen;
		clen += eclen;
		continue;

		/* Un-do work done so far */
fail:		while (i-- > 0) {
			structs_free(etype, NULL,
			    (char *)ary->elems + (i * etype->size));
		}
		FREE(mtype, ary->elems);
		return (-1);
	}

	/* Done */
	return (clen);
}

void
structs_array_free(const struct structs_type *type, void *data)
{
	const struct structs_type *const etype = type->args[0].v;
	const char *mtype = type->args[1].v;
	struct structs_array *const ary = data;
	u_int i;

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY)
		return;

	/* Free individual elements */
	for (i = 0; i < ary->length; i++)
		(*etype->uninit)(etype, (char *)ary->elems + (i * etype->size));

	/* Free array itself */
	FREE(mtype, ary->elems);
	memset(ary, 0, sizeof(*ary));
}

int
structs_array_length(const struct structs_type *type,
	const char *name, const void *data)
{
	const struct structs_array *ary = data;

	/* Find array */
	if ((type = structs_find(type, name, (const void **)&ary, 0)) == NULL)
		return (-1);

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Return length */
	return (ary->length);
}

int
structs_array_reset(const struct structs_type *type,
	const char *name, void *data)
{
	/* Find array */
	if ((type = structs_find(type, name, (const void **)&data, 1)) == NULL)
		return (-1);

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Free it, which resets it as well */
	structs_array_free(type, data);
	return (0);
}

int
structs_array_insert(const struct structs_type *type,
	const char *name, u_int index, void *data)
{
	struct structs_array *ary = data;
	const struct structs_type *etype;
	const char *mtype;
	void *mem;

	/* Find array */
	if ((type = structs_find(type, name, (const void **)&ary, 0)) == NULL)
		return (-1);
	etype = type->args[0].v;
	mtype = type->args[1].v;

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Check index */
	if (index > ary->length) {
		errno = EDOM;
		return (-1);
	}

	/* Reallocate array, leaving room for new element and a shift */
	if ((mem = REALLOC(mtype,
	    ary->elems, (ary->length + 2) * etype->size)) == NULL)
		return (-1);
	ary->elems = mem;

	/* Initialize new element; we'll move it into place later */
	if ((*etype->init)(etype,
	    (char *)ary->elems + (ary->length + 1) * etype->size) == -1)
		return (-1);

	/* Shift array over by one and move new element into place */
	memmove((char *)ary->elems + (index + 1) * etype->size,
	    (char *)ary->elems + index * etype->size,
	    (ary->length - index) * etype->size);
	memcpy((char *)ary->elems + index * etype->size,
	    (char *)ary->elems + (ary->length + 1) * etype->size, etype->size);
	ary->length++;
	return (0);
}

int
structs_array_delete(const struct structs_type *type,
	const char *name, u_int index, void *data)
{
	const struct structs_type *etype;
	struct structs_array *ary = data;

	/* Find array */
	if ((type = structs_find(type, name, (const void **)&ary, 0)) == NULL)
		return (-1);
	etype = type->args[0].v;

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Check index */
	if (index >= ary->length) {
		errno = EDOM;
		return (-1);
	}

	/* Free element */
	(*etype->uninit)(etype, (char *)ary->elems + (index * etype->size));

	/* Shift array */
	memmove((char *)ary->elems + index * etype->size,
	    (char *)ary->elems + (index + 1) * etype->size,
	    (--ary->length - index) * etype->size);
	return (0);
}

/*
 * Given a name, prepare all arrays in the name path for being
 * set with a new element.
 *
 * This means that any array index equal to the length of the array
 * causes a new element to be appended to the array rather than
 * an element not found error, and that any array index equal to
 * zero means to reset the entire array to be empty.
 *
 * This allows an entire array to be set by setting each element
 * in order, as long as this function is called with the name of
 * the element before each set operation.
 */
int
structs_array_prep(const struct structs_type *type,
	const char *name, void *data)
{
	char *nbuf;
	char *s;

	/* Skip if name cannot be an array element name */
	if (name == NULL || *name == '\0')
		return (0);

	/* Copy name into writable buffer */
	if ((nbuf = MALLOC(TYPED_MEM_TEMP, strlen(name) + 2)) == NULL)
		return (-1);
	nbuf[0] = '*';					/* for debugging */
	strcpy(nbuf + 1, name);

	/* Check all array element names in name path */
	for (s = nbuf; s != NULL; s = strchr(s + 1, STRUCTS_SEPARATOR)) {
		const struct structs_type *atype;
		struct structs_array *ary = data;
		u_long index;
		char *eptr;
		char *t;
		char ch;

		/* Go to next descendant node using prefix of name */
		ch = *s;
		*s = '\0';
		if ((atype = structs_find(type,
		    nbuf + (s != nbuf), (const void **)&ary, 1)) == NULL) {
			FREE(TYPED_MEM_TEMP, nbuf);
			return (-1);
		}
		*s = ch;

		/* If not array type, continue */
		if (atype->tclass != STRUCTS_TYPE_ARRAY)
			continue;

		/* Get array index */
		if ((t = strchr(s + 1, STRUCTS_SEPARATOR)) != NULL)
			*t = '\0';
		index = strtoul(s + 1, &eptr, 10);
		if (eptr == s + 1 || *eptr != '\0') {
			errno = ENOENT;
			FREE(TYPED_MEM_TEMP, nbuf);
			return (-1);
		}

		/* If setting an element that already exists, accept */
		if (index < ary->length)
			goto next;

		/* Must be setting the next new item in the array */
		if (index != ary->length) {
			errno = ENOENT;
			FREE(TYPED_MEM_TEMP, nbuf);
			return (-1);
		}

		/* Add new item; it will be in an initialized state */
		if (structs_array_insert(atype, NULL, ary->length, ary) == -1) {
			FREE(TYPED_MEM_TEMP, nbuf);
			return (-1);
		}

next:
		/* Repair name buffer for next time */
		if (t != NULL)
			*t = STRUCTS_SEPARATOR;
	}

	/* Done */
	FREE(TYPED_MEM_TEMP, nbuf);
	return (0);
}


/*
 * Given a name, allocate n items to a structs array.  If array is already
 * allocated it will be grown to nitems, if smaller than items the trailing
 * items will be truncated.
 */
int
structs_array_setsize(const struct structs_type *type,
		      const char *name, u_int nitems, void *data, int do_init)
{
	struct structs_array *ary = data;
	const struct structs_type *etype;
	const char *mem_type;
	void *mem;
	int idx;

	/* Find array */
	if ((type = structs_find(type, name, (const void **)&ary, 0)) == NULL)
		return (-1);
	etype = type->args[0].v;
	mem_type = type->args[1].v;

	/* Make sure it's really an array type */
	if (type->tclass != STRUCTS_TYPE_ARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Check size for negative and current (noop) */
	if (nitems < 0) {
		errno = EDOM;
		return (-1);
	}
	if (nitems == ary->length) {
		return (0);
	}
	/* Check size for 0 - free elements */
	if (nitems == 0) {
		FREE(mem_type, ary->elems);
		ary->elems = NULL;
		ary->length = 0;
		return(0);
	}

	/* Reallocate array, leaving room for new element and a shift */
	if ((mem = REALLOC(mem_type,
	    ary->elems, nitems * etype->size)) == NULL)
		return (-1);

	/* Zero out any new memory */
	if (nitems > ary->length) {
		memset(((char *)mem) + (ary->length * etype->size), 0, 
		       (nitems - ary->length) * etype->size);
	}
	
	/* Do initialization */
	if (do_init) {
	   for (idx = ary->length; idx < nitems; idx++) {
	      /* Initialize new element */
	      if ((*etype->init)(etype,
			 (char *)mem + (idx * etype->size)) == -1) {
		 return (-1);
	      }
	   }
	}
	ary->elems = mem;
	ary->length = nitems;
	return(0);
}


/*********************************************************************
			FIXEDARRAY TYPES
*********************************************************************/

int
structs_fixedarray_init(const struct structs_type *type, void *data)
{
	const struct structs_type *const etype = type->args[0].v;
	const u_int length = type->args[2].i;
	u_int i;

	for (i = 0; i < length; i++) {
		if ((*etype->init)(etype,
		    (char *)data + (i * etype->size)) == -1) {
			    while (i-- > 0) {
				    (*etype->uninit)(etype,
					(char *)data + (i * etype->size));
			    }
			    return (-1);
		}
	}
	return (0);
}

int
structs_fixedarray_copy(const struct structs_type *type,
	const void *from, void *to)
{
	const struct structs_type *const etype = type->args[0].v;
	const u_int length = type->args[2].i;
	u_int i;

	/* Make sure it's really a fixedarray type */
	if (type->tclass != STRUCTS_TYPE_FIXEDARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Copy elements into it */
	for (i = 0; i < length; i++) {
		const void *const from_elem = (char *)from + (i * etype->size);
		void *const to_elem = (char *)to + (i * etype->size);

		if ((*etype->copy)(etype, from_elem, to_elem) == -1)
			break;
	}

	/* If there was a failure, undo the half completed job */
	if (i < length) {
		while (i-- > 0)
			(*etype->uninit)(etype, (char *)to + (i * etype->size));
		return (-1);
	}

	/* Done */
	return (0);
}

int
structs_fixedarray_equal(const struct structs_type *type,
	const void *v1, const void *v2)
{
	const struct structs_type *const etype = type->args[0].v;
	const u_int length = type->args[2].i;
	u_int i;

	/* Make sure it's really a fixedarray type */
	if (type->tclass != STRUCTS_TYPE_FIXEDARRAY)
		return (0);

	/* Compare individual elements */
	for (i = 0;
	    i < length && (*etype->equal)(etype,
		(char *)v1 + (i * etype->size), (char *)v2 + (i * etype->size));
	    i++);
	return (i == length);
}

int
structs_fixedarray_encode(const struct structs_type *type, const char *mtype,
	struct structs_data *code, const void *data)
{
	const struct structs_type *const etype = type->args[0].v;
	const u_int length = type->args[2].i;
	const u_int bitslen = NUM_BYTES(length);
	struct structs_data *ecodes;
	u_char *bits;
	void *delem;
	u_int tlen;
	int r = -1;
	u_int i;

	/* Make sure it's really a fixedarray type */
	if (type->tclass != STRUCTS_TYPE_FIXEDARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Get the default value for an element */
	if ((delem = MALLOC(TYPED_MEM_TEMP, etype->size)) == NULL)
		return (-1);
	if (structs_init(etype, NULL, delem) == -1)
		goto fail1;

	/* Create bit array. Each bit indicates an element that is present. */
	if ((bits = MALLOC(TYPED_MEM_TEMP, bitslen)) == NULL)
		goto fail2;
	memset(bits, 0, bitslen);
	tlen = bitslen;

	/* Create array of individual encodings, one per element */
	if ((ecodes = MALLOC(TYPED_MEM_TEMP, length * sizeof(*ecodes))) == NULL)
		goto fail3;
	for (i = 0; i < length; i++) {
		const void *const elem = (char *)data + (i * etype->size);
		struct structs_data *const ecode = &ecodes[i];

		/* Check for default value, leave out if same as */
		if ((*etype->equal)(etype, elem, delem) == 1) {
			memset(ecode, 0, sizeof(*ecode));
			continue;
		}
		bits[i / 8] |= (1 << (i % 8));

		/* Encode element */
		if ((*etype->encode)(etype, TYPED_MEM_TEMP, ecode, elem) == -1)
			goto fail4;
		tlen += ecode->length;
	}

	/* Allocate final encoded region */
	if ((code->data = MALLOC(mtype, tlen)) == NULL)
		goto fail4;

	/* Copy bits array */
	memcpy(code->data, bits, bitslen);
	code->length = bitslen;

	/* Copy encoded elements */
	for (i = 0; i < length; i++) {
		struct structs_data *const ecode = &ecodes[i];

		memcpy(code->data + code->length, ecode->data, ecode->length);
		code->length += ecode->length;
	}

	/* OK */
	r = 0;

	/* Clean up and exit */
fail4:	while (i-- > 0)
		FREE(TYPED_MEM_TEMP, ecodes[i].data);
	FREE(TYPED_MEM_TEMP, ecodes);
fail3:	FREE(TYPED_MEM_TEMP, bits);
fail2:	structs_free(etype, NULL, delem);
fail1:	FREE(TYPED_MEM_TEMP, delem);
	return (r);
}

int
structs_fixedarray_decode(const struct structs_type *type, const u_char *code,
	size_t cmax, void *data, char *ebuf, size_t emax)
{
	const struct structs_type *const etype = type->args[0].v;
	const u_int length = type->args[2].i;
	const u_int bitslen = NUM_BYTES(length);
	const u_char *bits;
	int clen = 0;
	u_int i;

	/* Make sure it's really a fixedarray type */
	if (type->tclass != STRUCTS_TYPE_FIXEDARRAY) {
		errno = EINVAL;
		return (-1);
	}

	/* Get bits array */
	if (cmax < bitslen) {
		strlcpy(ebuf, "encoded array is truncated", emax);
		errno = EINVAL;
		return (-1);
	}
	bits = code;
	code += bitslen;
	cmax -= bitslen;
	clen += bitslen;

	/* Decode elements */
	for (i = 0; i < length; i++) {
		void *const edata = (char *)data + (i * etype->size);
		int eclen;

		/* If element not present, assign it the default value */
		if ((bits[i / 8] & (1 << (i % 8))) == 0) {
			if (structs_init(etype, NULL, edata) == -1)
				goto fail;
			continue;
		}

		/* Decode element */
		if ((eclen = (*etype->decode)(etype,
		    code, cmax, edata, ebuf, emax)) == -1)
			goto fail;

		/* Go to next encoded element */
		code += eclen;
		cmax -= eclen;
		clen += eclen;
		continue;

		/* Un-do work done so far */
fail:		while (i-- > 0) {
			structs_free(etype, NULL,
			    (char *)data + (i * etype->size));
		}
		return (-1);
	}

	/* Done */
	return (clen);
}

void
structs_fixedarray_free(const struct structs_type *type, void *data)
{
	const struct structs_type *const etype = type->args[0].v;
	const u_int length = type->args[2].i;
	u_int i;

	/* Make sure it's really a fixedarray type */
	if (type->tclass != STRUCTS_TYPE_FIXEDARRAY)
		return;

	/* Free elements */
	for (i = 0; i < length; i++)
		(*etype->uninit)(etype, (char *)data + (i * etype->size));
}

